### DISCUSSION TITLE
Adding M365Group to SharePoint Group

### DISCUSSION THREAD

**Discussion by pratheep-r** [pnp/powershell#3507] [Q&A]
I have a M365 group, which I need to add to SharePoint's member group when I create the site as a part of the script. Please see the error attached. 
![2023-10-16 17_19_43-Create-JDRProjectSite-Solution ps1 - JSON-Learn - Visual Studio Code](https://github.com/pnp/powershell/assets/56027088/f7a7e209-f4f7-448e-a6c1-bac7ca7ed1cd)

If I provide the full name of the SharePoint member group in "Add-PnPGroupMember" it works. 
Please advise/help.

**Comment by AndersRask** [comment]
I tried the above code and it works fine when Get-PnPGroup -AssociatedMemberGroup only returns one group.

If you have more than one group returned in line 23 PowerShell will give you an array of groups, not the group itself (this is a standard PowerShell thing).

Either iterate over groups and add them one at the time, or try with $SPMemberGroup[0]

**Reply by pratheep-r** [reply]
Yes that's correct. Get-PnPGroup -AssociatedMemberGroup only returns one member group (if you see the error message group is called "JDR Testing Script Check Members")

However if I add the group, throws the error.
Add-PnPGroupMember -Identity $SPMembersGroup -LoginName $M365PermssionSiteMembers

If I type in the Members group name like below, it works. 
Add-PnPGroupMember -Identity "JDR Testing Script Check Members" -LoginName $M365PermssionSiteMembers

**Reply by AndersRask** [reply]
could you check your object type on $SPMembersGroup using .GetType() please

**Reply by pratheep-r** [reply]
![image](https://github.com/pnp/powershell/assets/56027088/3c438557-dfb8-49b2-9f36-a17df6af8d14)

**Reply by AndersRask** [reply]
yup... that doesnt look right. It should say this:
$x = get-pnPGroup -AssociatedMemberGroup
$x.GetType()
```

IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     False    Group                                    Microsoft.SharePoint.Client.Principal
```

**Reply by AndersRask** [reply]
you can also see the full namespace of type using
```

$x | gm
   TypeName: Microsoft.SharePoint.Client.Group
```
Somehow your object has become a PSObject.

Also plz do a `$psversiontable` to see what version you are on

**Reply by pratheep-r** [reply]
Yes, doesn't look right. 

Full namespace;
![image](https://github.com/pnp/powershell/assets/56027088/1bd8fce5-10f0-4cdb-9043-2c1b33347c92)

Version;
![image](https://github.com/pnp/powershell/assets/56027088/caee47e0-3851-4532-b2b2-4d68b541c63e)

PnP Module;
![image](https://github.com/pnp/powershell/assets/56027088/0c0799f7-037a-4abc-99ac-d6d55244ae08)

Please get back if you need anything else. 
Thanks for helping.

**Reply by AndersRask** [reply]
Weird. It looks like you are getting SiteGroups instead of Groups (Get-PnPSiteGroup).

Could you post the code in its entirety (removing any real urls and confidential stuff).

You can paste it in and use the <> (CTRL+e) to format as code

**Reply by pratheep-r** [reply]
Clear-Host

Write-Host "Creating JDR Project Site" -ForegroundColor Green

$Title = "JDR SP Site Test"
#SiteAlias parameter to specify different alias for the Site URL
$SiteAlias = "JDR-SPS"
#To validation check 
$SiteUrl ="https://domain.sharepoint.com/sites/$SiteAlias"
#M365 Primary Email Address
$Alias = "JDR-SPS"
$Owners = "xxxxxxxxxxxxxxxxx"
$M365PermssionSiteMembers = "c:0o.c|federateddirectoryclaimprovider|GUID"
$JDRPnPProvisioningFile = "C:\xxxxxxx\xxxxx\JDR-PnP-Provisioning-File-Old.xml"

$Cred = Get-Credential
$AdminSiteURL = "https://domain-admin.sharepoint.com"
  
try {
    #Connect to SharePoint Tenant Admin Site
    $Connection = Connect-PnPOnline -Url $AdminSiteURL -Credentials $Cred -ReturnConnection
    $Site = Get-PnPTenantSite -Connection $Connection| Where-Object {$_.Url -eq $SiteUrl}

    if ($null -eq $Site) {
        #"New-PnPSite" doesn't set locale, timezone, time format - no parameters available.
        #Use "New-PnPTenantSite" if you are not applying template site
        $NewSite = New-PnPSite -Connection $Connection -Type TeamSite -Title $Title -SiteAlias $SiteAlias -Owners $Owners -Alias $Alias
        Write-Host "New JDR Site Collection $NewSite Created Successfully!" -ForegroundColor Green

        Add-Permission
    }
    else {
        write-host "Site $($SiteURL) exists already!" -foregroundcolor Yellow
        Write-Host "Exiting the Script" -ForegroundColor Yellow
        Start-Sleep 3
        Exit
    }
}
catch {
    write-host "Site creation Error: $($_.Exception.Message)" -foregroundcolor Red
    Write-Host "Exiting the Script" -ForegroundColor Yellow
    Start-Sleep 3
    Exit
}

function Add-Permission {
    try {
        Write-Host "`n"
        Write-Host "Adding M365 Permission Group" 
        Connect-PnPOnline -Url $NewSite -Credentials $Cred
        $SPMembersGroup = Get-PnPGroup -AssociatedMemberGroup
        Add-PnPGroupMember -Identity $SPMembersGroup -LoginName $M365PermssionSiteMembers
    }
    catch {
        write-host "Adding permission group Error: $($_.Exception.Message)" -foregroundcolor Red
    }
}

Write-Host "`n"
Write-Host "Pausing while site $NewSite is being provisioned" -ForegroundColor Green
Start-Sleep -Seconds 60
Write-Host "Site provisioning completed" -ForegroundColor Green

Add-SiteTemplate
function Add-SiteTemplate {
    try {
        Write-Host "`n"
        Write-Host "Applying JDR Project Site Template" -ForegroundColor Green
        #$JDRPnPProvisioningFile = $PSScriptRoot + "\Solutions" + "\JDR-PnP-Provisioning-File.xml"
        Connect-PnPOnline -url $NewSite -Credentials $Cred
        Invoke-PnPSiteTemplate -Path $JDRPnPProvisioningFile
    }
    catch {
        Write-host "Applying Site Template Error: $($_.Exception.Message)" -foregroundcolor Red
    }
}

``

**Reply by pratheep-r** [reply]
Apologies code formatting is not working as expected - perhaps I am missing something.

Another note, I am getting another problem. Site is not getting created, getting this error.
![image](https://github.com/pnp/powershell/assets/56027088/427d507f-e5f6-4438-a512-dc46c5cd1c14)
I am sure this part worked previously.

**Reply by AndersRask** [reply]
You are switching between connections using a connection parameter for one and not for the other.
This could be the issue, but not sure: try and use -ReturnConnection in your Add-Permission function and assign it to a $siteConn variable. Then use that connection whenever you are accessing the site rather than tenant.

And of course none of it will work if you are not creating a site... Try adding `-ValidateConnection `to your `connect-pnponline `to see if you actually are getting a connection

A second thing: when using try/catch you need to either enforce breaking errors using` $ErrorActionPreference = "STOP"` in the top of the script, or `-ErrorAction Stop` in the cmdlets being called inside the `try`. The reason for this is, that many errors in PowerShell are not terminating errors, so they will fail but not terminate the script, and hence not go into the `catch `part.

**Reply by pratheep-r** [reply]
Clear-Host

Write-Host "Creating JDR Project Site" -ForegroundColor Green

$Title = "JDR SP Site Test"
#SiteAlias parameter to specify different alias for the Site URL
$SiteAlias = "JDR-SPS"
#To validation check 
$SiteUrl ="https://domain.sharepoint.com/sites/$SiteAlias"
#M365 Primary Email Address
$Alias = "JDR-SPS"
$Owners = "xxxxxxxxxxxxxxxxxxxx"
$M365PermissionSiteMembers = "c:0o.c|federateddirectoryclaimprovider|GUID"
$JDRPnPProvisioningFile = "C:\xxxxxxxx\JDR\Solutions\JDR-PnP-Provisioning-File-Old.xml"

$Cred = Get-Credential
$AdminSiteURL = "https://domain-admin.sharepoint.com"
  
try {
    #Connect to SharePoint Tenant Admin Site
    $TenantConnection = Connect-PnPOnline -Url $AdminSiteURL -Credentials $Cred -ReturnConnection -ValidateConnection -ErrorAction Stop
    $Site = Get-PnPTenantSite -Connection $TenantConnection | Where-Object {$_.Url -eq $SiteUrl}

    if ($null -eq $Site) {
        #"New-PnPSite" doesn't set locale, timezone, time format - no parameters available.
        #Use "New-PnPTenantSite" if you are not applying template site
        $NewSite = New-PnPSite -Connection $TenantConnection -Type TeamSite -Title $Title -SiteAlias $SiteAlias -Owners $Owners -Alias $Alias -ErrorAction Stop
        Write-Host "New JDR Site Collection $NewSite Created Successfully!" -ForegroundColor Green

        Add-Permission
    }
    else {
        write-host "Site $($SiteURL) exists already!" -foregroundcolor Yellow
        Write-Host "Exiting the Script" -ForegroundColor Yellow
        Start-Sleep 3
        Exit
    }
}
catch {
    write-host "Site creation Error: $($_.Exception.Message)" -foregroundcolor Red
    Write-Host "Exiting the Script" -ForegroundColor Yellow
    Start-Sleep 3
    Exit
}

function Add-Permission {
    try {
        Write-Host "`n"
        Write-Host "Adding M365 Permission Group" -ForegroundColor Green
        $SiteConnection = Connect-PnPOnline -Url $NewSite -Credentials $Cred -ReturnConnection
        $SPMembersGroup = Get-PnPGroup -AssociatedMemberGroup -Connection $SiteConnection
        Add-PnPGroupMember -Connection $SiteConnection -Identity $SPMembersGroup -LoginName $M365PermissionSiteMembers
    }
    catch {
        write-host "Adding permission group Error: $($_.Exception.Message)" -foregroundcolor Red
    }
}

Write-Host "`n"
Write-Host "Pausing while site $NewSite is being provisioned" -ForegroundColor Green
Start-Sleep -Seconds 60
Write-Host "Site provisioning completed" -ForegroundColor Green

Add-SiteTemplate
function Add-SiteTemplate {
    try {
        Write-Host "`n"
        Write-Host "Applying JDR Project Site Template" -ForegroundColor Green
        #$JDRPnPProvisioningFile = $PSScriptRoot + "\Solutions" + "\JDR-PnP-Provisioning-File.xml"
        Connect-PnPOnline -url $NewSite -Connection $SiteConnection
        Invoke-PnPSiteTemplate -Connection $SiteConnection -Path $JDRPnPProvisioningFile
    }
    catch {
        Write-host "Applying Site Template Error: $($_.Exception.Message)" -foregroundcolor Red
    }
}``

**Reply by pratheep-r** [reply]
Thank you!
Added -validateConnection
Added -ErrorAction stop at tenant connection and site creation
Used -ReturnConnection for site and used the -Connection to subsequent codes require site connection.

Unfortunately still seeing the error;
![image](https://github.com/pnp/powershell/assets/56027088/c22cdc11-6b4a-43f1-846e-01bf38c6f1fb)

**Reply by AndersRask** [reply]
Try and enable tracing and paste log here (removing tenant relevant or sensitive info)

Set-PnPTraceLog -On -LogFile traceoutput.txt

**Reply by pratheep-r** [reply]
I tried like this just above the first try block. Sadly it is not logging outputs. 
Set-PnPTraceLog -On -LogFile "C:\Users\MC0147\xxxxx\Scripts\JDR\traceoutput.txt" -Level Debug

**Reply by AndersRask** [reply]
Don't set debug, then you will only get debug or higher. If you set nothing it should give Informational info level

**Reply by pratheep-r** [reply]
It is not outputting anything with or without debug.

**Comment by pratheep-r** [comment]
I tried "New-PnPTenantSite" the sites creates okay. However, this would not work as I cannot create Modern Team Site with M365 Group. 

Line 33 - $NewSite is not getting the new site object, therefore I am getting "Cannot bind argument to 'Url' because it is null". I am connecting to the $NewSite down the line to add permission and template. 

Line 22 - Trace log is on, but not doing anything. 
![image](https://github.com/pnp/powershell/assets/56027088/a485fc11-2e8d-4d5a-8205-83efdad2dfb7)

I have made some tweaks to script, please advise if you need the full script.

**Reply by AndersRask** [reply]
We will need the full stack trace to know where exactly it fails. 
In the catch try to write out something like below to get all info:

`Write-Error ("Error: [{0}]" -f ($_ | fl * -force | out-string))`

That should give us the whole error object with line numbers and stack trace.

Also there must be something changed, as this was not the original error you reported...

**Reply by pratheep-r** [reply]
I had to admit, I was running some testing for the proof of concept script with simple tests without validations, try, catch etc. 
Used "New-PnPSite" and then added permission group with few lines, where I stumbled on adding permission group to members group error, I reported first.  Site was created fine then. Since I expanded the script, "New-PnPSite" is not creating the site anymore. 

I am getting quite a lot of output with your new write-error - I noticed some quid, trace ID etc, is it safe to post here please?

**Reply by AndersRask** [reply]
Usually stack traces does not contain anything unsafe. You still have to check if there are domains in the code that you do not want.
Insert the output in full, select everything and use the <> button or CTRL+E to format as code - else it is quite unreadable.

Regarding site creation, the -Wait parameter ensures that all artifacts are created behind the scenes before it returns the result. It is a lot slower, but especially for debugging it ensures that you are not trying to call something before it is done being created.

**Reply by pratheep-r** [reply]
Site creation Error: AADSTS50076: Due to a configuration change made by your administrator, or because you moved to a new location, you must use multi-factor authentication to access '00000003-0000-0ff1-ce00-000000000000'.
Trace ID: 1de464c7-6b79-4c4f-96af-28185c793f00
Correlation ID: 9fcedeab-c3bb-4184-9cb0-28b859fd33f0
Timestamp: 2023-10-19 09:54:56Z
Write-Error: Error: [   
PSMessageDetails      : 
Exception             : MSAL.NetCore.4.50.0.0.MsalUiRequiredException: 
                                ErrorCode: invalid_grant
                        Microsoft.Identity.Client.MsalUiRequiredException: AADSTS50076: Due to a configuration change made by your administrator, or because you moved to a new location, you must use multi-factor authentication to access '00000003-0000-0ff1-ce00-000000000000'.
                        Trace ID: 1de464c7-6b79-4c4f-96af-28185c793f00
                        Correlation ID: 9fcedeab-c3bb-4184-9cb0-28b859fd33f0
                        Timestamp: 2023-10-19 09:54:56Z
                           at Microsoft.Identity.Client.OAuth2.OAuth2Client.ThrowServerException(HttpResponse response, RequestContext requestContext)
                           at Microsoft.Identity.Client.OAuth2.OAuth2Client.CreateResponse[T](HttpResponse response, RequestContext requestContext)
                           at Microsoft.Identity.Client.OAuth2.OAuth2Client.ExecuteRequestAsync[T](Uri endPoint, HttpMethod method, RequestContext requestContext, Boolean expectErrorsOn200OK, Boolean addCommonHeaders, Func`2 onBeforePostRequestData)
                           at Microsoft.Identity.Client.OAuth2.OAuth2Client.GetTokenAsync(Uri endPoint, RequestContext requestContext, Boolean addCommonHeaders, Func`2 onBeforePostRequestHandler)
                           at Microsoft.Identity.Client.OAuth2.TokenClient.SendHttpAndClearTelemetryAsync(String tokenEndpoint, ILoggerAdapter logger)
                           at Microsoft.Identity.Client.OAuth2.TokenClient.SendHttpAndClearTelemetryAsync(String tokenEndpoint, ILoggerAdapter logger)
                           at Microsoft.Identity.Client.OAuth2.TokenClient.SendTokenRequestAsync(IDictionary`2 additionalBodyParameters, String scopeOverride, String tokenEndpointOverride, CancellationToken cancellationToken)
                           at Microsoft.Identity.Client.Internal.Requests.UsernamePasswordRequest.GetTokenResponseAsync(CancellationToken cancellationToken)
                           at Microsoft.Identity.Client.Internal.Requests.UsernamePasswordRequest.ExecuteAsync(CancellationToken cancellationToken)
                           at Microsoft.Identity.Client.Internal.Requests.RequestBase.RunAsync(CancellationToken cancellationToken)
                           at Microsoft.Identity.Client.ApiConfig.Executors.PublicClientExecutor.ExecuteAsync(AcquireTokenCommonParameters commonParameters, AcquireTokenByUsernamePasswordParameters usernamePasswordParameters, CancellationToken cancellationToken)
                           at PnP.Framework.AuthenticationManager.GetContextAsync(String siteUrl, CancellationToken cancellationToken)
                           at PnP.Framework.AuthenticationManager.GetContext(String siteUrl)
                           at PnP.PowerShell.Commands.Base.PnPConnection.CreateWithCredentials(Cmdlet cmdlet, Uri url, PSCredential credentials, Boolean currentCredentials, String tenantAdminUrl, AzureEnvironment azureEnvironment, String clientId, String redirectUrl, Boolean onPrem, InitializationType initializationType) in c:\build\src\Commands\Base\PnPConnection.cs:line 437
                           at PnP.PowerShell.Commands.Base.ConnectOnline.ConnectCredentials(PSCredential credentials, InitializationType initializationType) in c:\build\src\Commands\Base\ConnectOnline.cs:line 578
                           at PnP.PowerShell.Commands.Base.ConnectOnline.Connect(CancellationToken& cancellationToken) in c:\build\src\Commands\Base\ConnectOnline.cs:line 301
                           at PnP.PowerShell.Commands.Base.ConnectOnline.ProcessRecord() in c:\build\src\Commands\Base\ConnectOnline.cs:line 247
                           at System.Management.Automation.CommandProcessor.ProcessRecord()
                                StatusCode: 400 
                                ResponseBody: {"error":"invalid_grant","error_description":"AADSTS50076: Due to a configuration change made by your administrator, or because you moved to a new location, you must use multi-factor authentication to access '00000003-0000-0ff1-ce00-000000000000'.\r\nTrace ID: 1de464c7-6b79-4c4f-96af-28185c793f00\r\nCorrelation ID: 9fcedeab-c3bb-4184-9cb0-28b859fd33f0\r\nTimestamp: 2023-10-19 09:54:56Z","error_codes":[50076],"timestamp":"2023-10-19 09:54:56Z","trace_id":"1de464c7-6b79-4c4f-96af-28185c793f00","correlation_id":"9fcedeab-c3bb-4184-9cb0-28b859fd33f0","error_uri":"https://login.microsoftonline.com/error?code=50076","suberror":"basic_action"}
                                Headers: Cache-Control: no-store, no-cache
                        Pragma: no-cache
                        Strict-Transport-Security: max-age=31536000; includeSubDomains
                        X-Content-Type-Options: nosniff
                        P3P: CP="DSP CUR OTPi IND OTRi ONL FIN"
                        client-request-id: 9fcedeab-c3bb-4184-9cb0-28b859fd33f0
                        x-ms-request-id: 1de464c7-6b79-4c4f-96af-28185c793f00
                        x-ms-ests-server: 2.1.16522.6 - FRC ProdSlices
                        x-ms-clitelem: 1,50076,0,,
                        X-XSS-Protection: 0
                        Set-Cookie: fpc=Aul3pTtwfDJCgifAyUSmSR9l9v-HAQAAAPDzwtwOAAAA; expires=Sat, 18-Nov-2023 09:54:56 GMT; path=/; secure; HttpOnly; SameSite=None, x-ms-gateway-slice=estsfd; path=/; secure; samesite=none; httponly, stsservicecookie=estsfd; path=/; secure; samesite=none; httponly
                        Date: Thu, 19 Oct 2023 09:54:56 GMT

TargetObject          : 
CategoryInfo          : NotSpecified: (:) [Connect-PnPOnline], MsalUiRequiredException
FullyQualifiedErrorId : Microsoft.Identity.Client.MsalUiRequiredException,PnP.PowerShell.Commands.Base.ConnectOnline
ErrorDetails          :
InvocationInfo        : System.Management.Automation.InvocationInfo
ScriptStackTrace      : at <ScriptBlock>, C:\Users\MC0147\xxxxxxxx\Scripts\JDR\Create-JDRProjectSite.ps1: line 26
                        at <ScriptBlock>, <No file>: line 1
PipelineIterationInfo : {}

]``

**Reply by AndersRask** [reply]
looks like your -Credentials connection is not working because your organization has roled out requirement for MFA

**Due to a configuration change made by your administrator, or because you moved to a new location, you must use multi-factor authentication to access '00000003-0000-0ff1-ce00-000000000000**

Try to connect using -Interactive instead of credentials

**Reply by pratheep-r** [reply]
Like this in line 26? 
Do I also need to add -Interactive in line 32?
![image](https://github.com/pnp/powershell/assets/56027088/87c331bf-4e97-49ae-b18b-c68374ba34e3)

**Reply by AndersRask** [reply]
yes, everywhere where you used Credentials try instead with Interactive
Interactive uses modern authentication that will prompt in a browser modal window for login with MFA

**Reply by pratheep-r** [reply]
With the -Interactive 

Site creation Error: Object reference not set to an instance of an object.
Write-Error: Error: [
PSMessageDetails      : 
Exception             : System.NullReferenceException: Object reference not set to an instance of an object.
                           at PnP.PowerShell.Commands.NewSite.ExecuteCmdlet() in c:\build\src\Commands\Admin\NewSite.cs:line 133
                           at PnP.PowerShell.Commands.Base.PnPConnectedCmdlet.ProcessRecord() in c:\build\src\Commands\Base\PnPConnectedCmdlet.cs:line 53
TargetObject          : 
CategoryInfo          : WriteError: (:) [New-PnPSite], NullReferenceException
FullyQualifiedErrorId : EXCEPTION,PnP.PowerShell.Commands.NewSite
ErrorDetails          : Object reference not set to an instance of an object.
InvocationInfo        : System.Management.Automation.InvocationInfo
ScriptStackTrace      : at <ScriptBlock>, C:\Users\MC0147\xxxxxxxx\Scripts\JDR\Create-JDRProjectSite.ps1: line 32
                        at <ScriptBlock>, <No file>: line 1
PipelineIterationInfo : {}


]``

**Reply by AndersRask** [reply]
Ok, you will have to find out what of your parameters it is failing on.
Line 133 in the framework takes these parameters

`var returnedContext = Framework.Sites.SiteCollection.Create(ClientContext, creationInformation, noWait: !Wait, graphAccessToken: GraphAccessToken, azureEnvironment: Connection?.AzureEnvironment ?? Framework.AzureEnvironment.Production);`

I would venture that your connection/context is not valid.
Try and do a simple thing in the console, like fetching lists on the site to check if you have a valid connection.
Also try to create the site "by hand" in the console, where you type the parameters like Title, Url etc by hand to test where it fails.

**Reply by pratheep-r** [reply]
Apologies if I misunderstood, "create site 'by hand' in the console" - I am able to create site from admin portal. 

Although it error'ed when I use the "New-PnPSite"
![image](https://github.com/pnp/powershell/assets/56027088/03480f00-ffcd-47c7-826a-41aecbceec73)

Site is created when I use "New-PnPTenantSite"
![image](https://github.com/pnp/powershell/assets/56027088/76a615ec-0b1a-47b4-a7b1-66f2259ef768)
Again $NewSite does not have the created site object. However, site is actually created.
![image](https://github.com/pnp/powershell/assets/56027088/65471357-bca2-41f4-a06e-828a0521bbc3)
So, appears, connection works when I use "New-PnPTenantsite"

It is really odd, I am running out of option. I am going to update my PnP module - I will have to wait get temp local admin rights sorted first.

**Reply by AndersRask** [reply]
I tried the above code, and both ways of creating a site works fine for me

**Reply by pratheep-r** [reply]
I updated the PnP.PowerShell module to 2.2.0, but to no avail. 
Update-module threw an error something along the line of "was not installed by using Install-Module" although it was installed via "Install-Module".
I deleted the physical folder, used "Install-Module" scoping to all users. 
I don't suspect anything to do with this, but thinking out loud, anything appears unusual?
